segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+
+            // Evict chunks if required.
+            if (maxIndexedChunksPerSegment < segmentReadIndex.chunkIndex.size() + 1
+                    || maxIndexedChunks < totalChunkCount.get() + 1) {
+                evictChunks(streamSegmentName, 1);
+            }
+
+            segmentReadIndex.chunkIndex.put(startOffset,
+                    SegmentReadIndexEntry.builder()
+                            .chunkName(chunkName)
+                            .generation(new AtomicLong(currentGeneration.get()))
+                            .build());
+            segmentReadIndex.generation.set(currentGeneration.get());
+            totalChunkCount.incrementAndGet();
+        }
+    }
+
+    /**
+     * Updates read index for given segment with new entries.
+     *
+     * @param streamSegmentName
+     * @param newReadIndexEntries List of {@link ChunkNameOffsetPair} for new entries.
+     */
+    public void addIndexEntries(String streamSegmentName, List<ChunkNameOffsetPair> newReadIndexEntries) {
+        val segmentReadIndex = getSegmentReadIndex(streamSegmentName);
+        // Evict chunks if required.
+        if (maxIndexedChunksPerSegment < segmentReadIndex.chunkIndex.size() + newReadIndexEntries.size()
+                || maxIndexedChunks < totalChunkCount.get() + newReadIndexEntries.size()) {
+            evictChunks(streamSegmentName, newReadIndexEntries.size());
+        }
+
+        for (val entry : newReadIndexEntries) {
+            segmentReadIndex.chunkIndex.put(entry.getOffset(),
+                    SegmentReadIndexEntry.builder()
+                            .chunkName(entry.getChunkName())
+                            .generation(new AtomicLong(currentGeneration.get()))
+                            .build());
+            segmentReadIndex.generation.set(currentGeneration.get());
+
+        }
+        totalChunkCount.getAndAdd(newReadIndexEntries.size());
+    }
+
+    /**
+     * Removes the given segment from cache.
+     *
+     * @param streamSegmentName
+     */
+    public void remove(String streamSegmentName) {
+        val readIndex = segmentsToReadIndexMap.get(streamSegmentName);
+        if (null != readIndex) {
+            segmentsToReadIndexMap.remove(streamSegmentName);
+            totalChunkCount.getAndAdd(-1 * readIndex.chunkIndex.size());
+        }
+    }
+
+    /**
+     * Finds a chunk that is floor to the given offset.
+     *
+     * @param streamSegmentName Name of the segment.
+     * @param offset            Offset for which to search.
+     * @return
+     */
+    public ChunkNameOffsetPair findFloor(String streamSegmentName, long offset) {
+        val segmentReadIndex = getSegmentReadIndex(streamSegmentName);
+        if (segmentReadIndex.chunkIndex.size() > 0) {
+            val floorEntry = segmentReadIndex.chunkIndex.floorEntry(offset);
+            if (null != floorEntry) {
+                // mark with current generation
+                segmentReadIndex.generation.set(currentGeneration.get());
```

Comment:
Assuming that client is reading sequentially, it is likely that the same chunk will be read in subsequent calls. So we need to keep that entry in cache. Because this was a cache hit we need to mark it with current generation.